Seeding details EDM update

guilhermeAlmeida1 opened the pull request:
Similarly to #248 , updated the seeding specific data types (listed below) to match others.

- singlet (not actually being used anywhere in the code, might be worth revisiting
- doublet
- triplet
- lin_circle
- bound_track_parameters

krasznaa commented on core/include/traccc/seeding/detail/lin_circle.hpp:
```diff
@@ -48,37 +48,10 @@ struct lin_circle {
-/// host code
-using lin_circle_container_view = container_view<unsigned int, lin_circle>;
+/// Declare all lin_circle container types
+using lin_circle_container_types = container_types<unsigned int, lin_circle>;
```
Let's not even declare this type. It doesn't seem to be used anywhere. (And the first argument of the template seems to be a dummy anyway.)

krasznaa commented on core/include/traccc/seeding/detail/singlet.hpp:
```diff
@@ -30,32 +30,10 @@ inline TRACCC_HOST_DEVICE bool operator!=(const sp_location& lhs,
-using singlet_collection = vector_t<sp_location>;
+/// Declare all singlet collection types
+using singlet_collection_types = collection_types<sp_location>;
 
-/// Convenience declaration for the singlet collection type to use in host code
-using host_singlet_collection = singlet_collection<vecmem::vector>;
+/// Declare all singlet container types
+using singlet_container_types = container_types<unsigned int, sp_location>;
```
Yeah, let's just drop these definitions. If they are not needed, let's not burden the compiler with them.

krasznaa commented on core/include/traccc/seeding/doublet_finding.hpp:
```diff
@@ -18,9 +18,9 @@ namespace traccc {
-              const sp_grid&, const sp_location&, const bool&)> {
+    : public algorithm<std::pair<doublet_collection_types::host,
+                                 lin_circle_collection_types::host>(
+          const sp_grid&, const sp_location&, const bool&)> {
```
Not for this PR, but wow... We really provide a boolean to the algorithm as a constant reference?!? :scream: We should really fix that in a separate PR.

krasznaa commented on device/common/include/traccc/seeding/device/impl/select_seeds.ipp:
```diff
@@ -81,14 +81,14 @@ void select_seeds(
     const unsigned int num_triplets_per_bin =
         triplet_device.get_headers().at(bin_idx).n_triplets;
-    const vecmem::device_vector<triplet> triplets_per_bin =
+    const vecmem::device_vector<const triplet> triplets_per_bin =
```
Instead of the explicit type name, which is correct, we should use either

```c++
triplet_collection_types::const_device
```

, or better yet:

```c++
triplet_container_types::item_vector::const_reference
```

(I hope I wrote the correct types. I didn't try to compile this of course.) This way if the underlying types change, as long as the type definitions are correctly defined, this will keep working.

guilhermeAlmeida1 replied:
`triplet_container_types::const_device::item_vector::const_reference`

I think this version is a bit overkill and goes against what we're using throughout all our device kernels. Isn't the point of having both the collection & container types declared to also be able to use them together coherently?

guilhermeAlmeida1 replied:
(so I changed it to `triplet_collection_types::const_device`)

krasznaa replied:
Fine...

krasznaa reviewed (changes requested):
Looks good. Though at one point we should really clean up how we pass arguments to the `traccc::device::` functions. Right now we pass some constant arguments by value, and some by const reference. We should just use one. (I think const reference would be better.)

guilhermeAlmeida1 commented:
@krasznaa Other than the requested changes, while I was looking at track_params_est I noticed that there we were still using some older version of container classes there, so I changed those as well and added to this PR

guilhermeAlmeida1 commented on core/include/traccc/edm/track_parameters.hpp:
```diff
@@ -98,19 +98,25 @@ inline bool operator==(const bound_track_parameters& lhs,
-/// in device code
-using device_bound_track_parameters_collection =
-    bound_track_parameters_collection<vecmem::device_vector>;
+// template <template <typename> class vector_t>
+// using bound_track_parameters_collection = vector_t<bound_track_parameters>;
+
+// /// Convenient declaration for the bound track parameters collection type to
+// use
+// /// in host code
+// using host_bound_track_parameters_collection =
+//     bound_track_parameters_collection<vecmem::vector>;
+
+// /// Convenient declaration for the bound track parameters collection type to
+// use
+// /// in device code
+// using device_bound_track_parameters_collection =
+//     bound_track_parameters_collection<vecmem::device_vector>;
```
whoops, missed this

krasznaa reviewed (approved):
Yeah, let's do it.
